### DISCUSSION TITLE
`azd` hooks vs CI/CD pipelines

### DISCUSSION THREAD

**Discussion by lechnerc77** [Azure/azure-dev#1613] [General]
Hi community, 

I have a point I would like to discuss: with the advent of the hooks in `azd` several new options to enhance the functionality of  `azd` are available. 

One interesting scenario was shown in the February community standup, where a service lifecycle event hook was leveraged to execute the tests defined in the app before doing the deployment. Thinking a bit about this scenario raises the question in my mind of what to put into the `azd` definition and what to put into a CI/CD pipeline. 
Although I like the spin of keeping things self-contained and having this step as part of the `azure.yaml`, I think I would have put that a s a dedicated step in a GitHub action as this is at least from my perspective more explicit while "hiding" this as part of the `azd` is more implicit. 

Although I think there is no "right" or "wrong" here, I would be very interested in hearing your opinions on this topic.

**Comment by rajyraman** [comment]
@jongio - Since there is no context passed to the hooks, what is the real advantage of using hooks vs doing this entirely in the pipeline/workflow?

**Comment by vhvb1989** [comment]
@rajyraman, using `hooks` would work both locally and from CI/CD.
The hooks also gets the env-vars from azd automatically.

**Reply by rajyraman** [reply]
@vhvb1989 - I am still using azd env get-values in hooks. Are you saying that I don't need to do that?

**Reply by vhvb1989** [reply]
correct. azd injects all the values from the .env file automatically.  You can remove the `azd env get-values`

**Reply by rajyraman** [reply]
@vhvb1989 - Tried it now and it works 😁. Didn't know that I didn't have to do azd env get-values at all. ~This resolves my [other question](https://github.com/Azure/azure-dev/discussions/2603) about hooks as well.~ Thank you Victor.

**Reply by rajyraman** [reply]
@vhvb1989 - Please document this feature. It doesn't seem to be mentioned in https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/manage-environment-variables or https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/azd-extensibility

**Reply by rajyraman** [reply]
@vhvb1989 - Please also clarify what this means for https://github.com/Azure/azure-dev/issues/1697

**Reply by vhvb1989** [reply]
Created: https://github.com/Azure/azure-dev/issues/2607 to track updating the hooks documentation. @wbreza @savannahostrowski @gracewilcox FYI

**Reply by vhvb1989** [reply]
@rajyraman , #1697 is about pulling the azd-env into your current terminal-session (regardless of running inside a hook).
You might want this when you are testing the script that is invoked by a hook without using azd.  For example, for script `foo.sh` like

```sh
echo $AZD_ENV_VAR
```

If it is set as azd-hook, azd will automatically inject the environment and you would see the expected output (when azd runs the hook).
But if you run the `foo.sh` manually, you wouldn't see the expected output because the azd-env is not set in your terminal.

**Reply by savannahostrowski** [reply]
Wrong Grace! cc: @gkulin for docs
